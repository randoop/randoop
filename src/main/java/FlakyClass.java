public class FlakyClass {
  public int flakyDefaultHashCode() {
    return super.hashCode();
  }
}
