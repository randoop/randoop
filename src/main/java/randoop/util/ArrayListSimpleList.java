package randoop.util;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class ArrayListSimpleList<T> extends SimpleList<T> implements Serializable {

  private static final long serialVersionUID = 9155161101212598259L;

  public final ArrayList<T> theList;

  public ArrayListSimpleList(ArrayList<T> list) {
    theList = new ArrayList<>(list);
  }

  public ArrayListSimpleList() {
    theList = new ArrayList<>();
  }

  public ArrayListSimpleList(int capacity) {
    theList = new ArrayList<>(capacity);
  }

  @Override
  public int size() {
    return theList.size();
  }

  @Override
  public T get(int index) {
    return theList.get(index);
  }

  @Override
  public SimpleList<T> getSublist(int index) {
    return this;
  }

  public boolean add(T element) {
    return theList.add(element);
  }

  @Override
  public List<T> toJDKList() {
    return new ArrayList<>(theList);
  }

  @Override
  public String toString() {
    return toJDKList().toString();
  }
}
