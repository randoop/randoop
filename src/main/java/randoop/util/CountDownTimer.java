package randoop.util;

public class CountDownTimer {

  private long totalTime;
  private long startTime;

  private CountDownTimer(long timeMillis) {
    this.totalTime = timeMillis;
    this.startTime = System.currentTimeMillis();
  }

  public boolean reachedZero() {
    return (elapsedTime() >= totalTime);
  }

  public long elapsedTime() {
    return System.currentTimeMillis() - this.startTime;
  }

  public long remainingTime() {
    long remainingTime = totalTime - elapsedTime();
    if (remainingTime < 0) {
      return 0;
    }
    return remainingTime;
  }

  public static CountDownTimer createAndStart(long totalTimeMillis) {
    return new CountDownTimer(totalTimeMillis);
  }

  @Override
  public String toString() {
    return "elapsed: " + elapsedTime() + ", remaining: " + remainingTime();
  }
}
