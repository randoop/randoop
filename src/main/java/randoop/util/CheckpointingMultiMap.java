package randoop.util;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * A MultiMap that supports checkpointing and restoring to a checkpoint (that is, undoing all
 * operations up to a checkpoint, also called a "mark").
 */
public class CheckpointingMultiMap<K, V> implements IMultiMap<K, V> {

  public static boolean verbose_log = false;

  private final Map<K, Set<V>> map;

  public final List<Integer> marks;

  private enum Ops {
    ADD,
    REMOVE
  }

  private final List<OpKeyVal> ops;

  private int steps;

  // A triple of an operation, a key, and a value
  private class OpKeyVal {
    final Ops op;
    final K key;
    final V val;

    OpKeyVal(final Ops op, final K key, final V val) {
      this.op = op;
      this.key = key;
      this.val = val;
    }
  }

  public CheckpointingMultiMap() {
    map = new LinkedHashMap<>();
    marks = new ArrayList<>();
    ops = new ArrayList<>();
    steps = 0;
  }

  @Override
  public boolean add(K key, V value) {
    if (verbose_log) {
      Log.logPrintf("ADD %s -> %s%n", key, value);
    }
    add_bare(key, value);
    ops.add(new OpKeyVal(Ops.ADD, key, value));
    steps++;
    return true;
  }

  private void add_bare(K key, V value) {
    if (key == null || value == null) {
      throw new IllegalArgumentException("args cannot be null.");
    }

    Set<V> values = map.computeIfAbsent(key, __ -> new LinkedHashSet<>(1));
    if (values.contains(value)) {
      throw new IllegalArgumentException("Mapping already present: " + key + " -> " + value);
    }
    values.add(value);
  }

  @Override
  public boolean remove(K key, V value) {
    if (verbose_log) {
      Log.logPrintf("REMOVE %s -> %s%n", key, value);
    }
    remove_bare(key, value);
    ops.add(new OpKeyVal(Ops.REMOVE, key, value));
    steps++;
    return true;
  }

  private void remove_bare(K key, V value) {
    if (key == null || value == null) {
      throw new IllegalArgumentException("args cannot be null.");
    }

    Set<V> values = map.get(key);
    if (values == null) {
      throw new IllegalArgumentException("Mapping not present: " + key + " -> " + value);
    }
    values.remove(value);

    // If no more mapping from key, remove key from map.
    if (values.isEmpty()) {
      map.remove(key);
    }
  }

  /** Checkpoint the state of the data structure, for use by {@link #undoToLastMark()}. */
  public void mark() {
    marks.add(steps);
    steps = 0;
  }

  /** Undo changes since the last call to {@link #mark()}. */
  public void undoToLastMark() {
    if (marks.isEmpty()) {
      throw new IllegalArgumentException("No marks.");
    }
    Log.logPrintf("marks: %s%n", marks);
    for (int i = 0; i < steps; i++) {
      undoLastOp();
    }
    steps = marks.remove(marks.size() - 1);
  }

  private void undoLastOp() {
    if (ops.isEmpty()) throw new IllegalStateException("ops empty.");
    OpKeyVal last = ops.remove(ops.size() - 1);
    Ops op = last.op;
    K key = last.key;
    V val = last.val;

    if (op == Ops.ADD) {
      // Remove the mapping.
      Log.logPrintf("REMOVE %s%n", key + " ->" + val);
      remove_bare(key, val);
    } else if (op == Ops.REMOVE) {
      // Add the mapping.
      Log.logPrintf("ADD %s -> %s%n", key, val);
      add_bare(key, val);
    } else {
      // Really, we should never get here.
      throw new IllegalStateException("Unhandled op: " + op);
    }
  }

  @Override
  public Set<V> getValues(K key) {
    if (key == null) throw new IllegalArgumentException("arg cannot be null.");
    return map.getOrDefault(key, Collections.emptySet());
  }

  public boolean containsKey(Object key) {
    if (key == null) throw new IllegalArgumentException("arg cannot be null.");
    return map.containsKey(key);
  }

  @Override
  public Set<K> keySet() {
    return map.keySet();
  }

  @Override
  public int size() {
    return map.size();
  }

  @Override
  public String toString() {
    return map.toString();
  }
}
