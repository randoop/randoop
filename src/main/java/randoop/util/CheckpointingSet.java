package randoop.util;

import java.util.Collection;
import java.util.Iterator;
import java.util.Set;

/**
 * A Set that supports settingcheckpoints (also called "marks") and restoring the data structure's
 * state to them.
 */
public class CheckpointingSet<E> implements Set<E> {

  // This uses a MultiMap just because that is an existing checkpointing data structure.
  // The value is always true in this mapping, never false.
  public final CheckpointingMultiMap<E, Boolean> map;

  public CheckpointingSet() {
    this.map = new CheckpointingMultiMap<>();
  }

  @Override
  public boolean add(E elt) {
    if (elt == null) throw new IllegalArgumentException("arg cannot be null.");
    if (contains(elt)) throw new IllegalArgumentException("set already contains elt " + elt);
    return map.add(elt, Boolean.TRUE);
  }

  @Override
  public boolean contains(Object elt) {
    if (elt == null) throw new IllegalArgumentException("arg cannot be null.");
    return map.containsKey(elt);
  }

  @Override
  public boolean remove(Object elt) {
    if (elt == null) {
      throw new IllegalArgumentException("arg cannot be null.");
    }

    @SuppressWarnings("unchecked")
    E eltCasted = (E) elt;
    return map.remove(eltCasted, Boolean.TRUE);
  }

  @Override
  public int size() {
    return map.size();
  }

  /** Checkpoint the state of the data structure, for use by {@link #undoToLastMark()}. */
  public void mark() {
    map.mark();
  }

  /** Undo changes since the last call to {@link #mark()}. */
  public void undoToLastMark() {
    map.undoToLastMark();
  }

  @Override
  public String toString() {
    return map.keySet().toString();
  }

  @Override
  public void clear() {
    throw new UnsupportedOperationException("not yet implemented");
  }

  @Override
  public boolean addAll(Collection<? extends E> c) {
    throw new UnsupportedOperationException("not yet implemented");
  }

  @Override
  public boolean removeAll(Collection<?> c) {
    throw new UnsupportedOperationException("not yet implemented");
  }

  @Override
  public boolean retainAll(Collection<?> c) {
    throw new UnsupportedOperationException("not yet implemented");
  }

  @Override
  public boolean containsAll(Collection<?> c) {
    throw new UnsupportedOperationException("not yet implemented");
  }

  @Override
  public <T> T[] toArray(T[] a) {
    throw new UnsupportedOperationException("not yet implemented");
  }

  @Override
  public Object[] toArray() {
    throw new UnsupportedOperationException("not yet implemented");
  }

  @Override
  public Iterator<E> iterator() {
    throw new UnsupportedOperationException("not yet implemented");
  }

  @Override
  public boolean isEmpty() {
    // return map.isEmpty();
    throw new UnsupportedOperationException("not yet implemented");
  }
}
