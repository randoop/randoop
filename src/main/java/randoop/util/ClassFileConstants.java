package randoop.util;

import java.io.IOException;
import java.io.InputStream;
import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.StringJoiner;
import java.util.TreeSet;
import org.apache.bcel.Const;
import org.apache.bcel.classfile.ClassParser;
import org.apache.bcel.classfile.Constant;
import org.apache.bcel.classfile.ConstantClass;
import org.apache.bcel.classfile.ConstantDouble;
import org.apache.bcel.classfile.ConstantFieldref;
import org.apache.bcel.classfile.ConstantFloat;
import org.apache.bcel.classfile.ConstantInteger;
import org.apache.bcel.classfile.ConstantInterfaceMethodref;
import org.apache.bcel.classfile.ConstantInvokeDynamic;
import org.apache.bcel.classfile.ConstantLong;
import org.apache.bcel.classfile.ConstantMethodHandle;
import org.apache.bcel.classfile.ConstantMethodType;
import org.apache.bcel.classfile.ConstantMethodref;
import org.apache.bcel.classfile.ConstantNameAndType;
import org.apache.bcel.classfile.ConstantPool;
import org.apache.bcel.classfile.ConstantString;
import org.apache.bcel.classfile.ConstantUtf8;
import org.apache.bcel.classfile.JavaClass;
import org.apache.bcel.classfile.Method;
import org.apache.bcel.generic.ClassGen;
import org.apache.bcel.generic.ConstantPoolGen;
import org.apache.bcel.generic.ConstantPushInstruction;
import org.apache.bcel.generic.Instruction;
import org.apache.bcel.generic.InstructionList;
import org.apache.bcel.generic.MethodGen;
import org.apache.bcel.util.ClassPath;
import org.checkerframework.checker.signature.qual.ClassGetName;
import randoop.operation.NonreceiverTerm;
import randoop.reflection.TypeNames;
import randoop.types.JavaTypes;

// Implementation notes:  All string, float, and double constants are in the
// the constant table.  Integer constants less that 64K are in the code.
// There are also special opcodes to push values from -1 to 5.  This code
// does not include them, but it would be easy to add them.  This code also
// does not include class literals as constants.
// It would be possible to determine the method with the constant if you
// wanted finer-grained information about where the constants were used.

/**
 * Reads literals from a class file, including from the constant pool and from bytecodes that take
 * immediate arguments.
 */
public class ClassFileConstants {

  // Some test values when this class file is used as input.
  // Byte, int, short, and char values are all stored in the .class file as int.
  static byte bb = 23;
  static double d = 35.3;
  static float f = 3.0f;
  static int ii = 20;
  static long ll = 200000;
  static short s = 32000;
  static char c = 'a';

  public static class ConstantSet {
    public @ClassGetName String classname;
    public Set<Integer> ints = new TreeSet<>();
    public Set<Long> longs = new TreeSet<>();
    public Set<Float> floats = new TreeSet<>();
    public Set<Double> doubles = new TreeSet<>();
    public Set<String> strings = new TreeSet<>();
    public Set<Class<?>> classes = new HashSet<>();

    @Override
    public String toString() {
      StringJoiner sb = new StringJoiner(randoop.Globals.lineSep);

      sb.add("START CLASSLITERALS for " + classname);
      for (int x : ints) {
        sb.add("int:" + x);
      }
      for (long x : longs) {
        sb.add("long:" + x);
      }
      for (float x : floats) {
        sb.add("float:" + x);
      }
      for (double x : doubles) {
        sb.add("double:" + x);
      }
      for (String x : strings) {
        sb.add("String:\"" + x + "\"");
      }
      for (Class<?> x : classes) {
        sb.add("Class:" + x);
      }
      sb.add("%nEND CLASSLITERALS for " + classname);

      return sb.toString();
    }
  }

  /**
   * A simple driver program that prints output literals file format.
   *
   * @param args the command line arguments
   * @throws IOException if an error occurs in writing the constants
   * @see randoop.reflection.LiteralFileReader
   */
  public static void main(String[] args) throws IOException {
    for (String classname : args) {
      System.out.println(getConstants(classname));
    }
  }

  /**
   * Returns all the constants found in the given class.
   *
   * @param classname the name of the type
   * @return the set of constants of the given type
   * @see #getConstants(String,ConstantSet)
   */
  public static ConstantSet getConstants(String classname) {
    ConstantSet result = new ConstantSet();
    getConstants(classname, result);
    return result;
  }

  /**
   * Adds all the constants found in the given class into the given ConstantSet, and returns it.
   *
   * @param classname the name of the type
   * @param result the set of constants to which constants are added
   * @return the set of constants with new constants of given type added
   * @see #getConstants(String)
   */
  public static ConstantSet getConstants(String classname, ConstantSet result) {

    String classfileBase = classname.replace('.', '/');
    ClassParser cp;
    JavaClass jc;
    try (InputStream is = ClassPath.SYSTEM_CLASS_PATH.getInputStream(classfileBase, ".class")) {
      cp = new ClassParser(is, classname);
      jc = cp.parse();
    } catch (java.io.IOException e) {
      throw new Error("IOException while reading '" + classname + "': " + e.getMessage());
    }
    @SuppressWarnings("signature") // BCEL's JavaClass is not annotated for the Signature Checker
    @ClassGetName String resultClassname = jc.getClassName();
    result.classname = resultClassname;

    // Get all of the constants from the pool
    ConstantPool constant_pool = jc.getConstantPool();
    for (Constant c : constant_pool.getConstantPool()) {
      // System.out.printf ("*Constant = %s%n", c);
      if (c == null
          || c instanceof ConstantClass
          || c instanceof ConstantFieldref
          || c instanceof ConstantInterfaceMethodref
          || c instanceof ConstantMethodref
          || c instanceof ConstantNameAndType
          || c instanceof ConstantMethodHandle
          || c instanceof ConstantMethodType
          || c instanceof ConstantInvokeDynamic
          || c instanceof ConstantUtf8) {
        continue;
      }
      if (c instanceof ConstantString) {
        result.strings.add((String) ((ConstantString) c).getConstantValue(constant_pool));
      } else if (c instanceof ConstantDouble) {
        result.doubles.add((Double) ((ConstantDouble) c).getConstantValue(constant_pool));
      } else if (c instanceof ConstantFloat) {
        result.floats.add((Float) ((ConstantFloat) c).getConstantValue(constant_pool));
      } else if (c instanceof ConstantInteger) {
        result.ints.add((Integer) ((ConstantInteger) c).getConstantValue(constant_pool));
      } else if (c instanceof ConstantLong) {
        result.longs.add((Long) ((ConstantLong) c).getConstantValue(constant_pool));
      } else {
        throw new RuntimeException("Unrecognized constant of type " + c.getClass() + ": " + c);
      }
    }

    ClassGen gen = new ClassGen(jc);
    ConstantPoolGen pool = gen.getConstantPool();

    // Process the code in each method looking for literals
    for (Method m : jc.getMethods()) {
      @SuppressWarnings("signature") // BCEL's JavaClass is not annotated for the Signature Checker
      MethodGen mg = new MethodGen(m, jc.getClassName(), pool);
      InstructionList il = mg.getInstructionList();
      if (il != null) {
        for (Instruction inst : il.getInstructions()) {
          switch (inst.getOpcode()) {

              // Compare two objects, no literals
            case Const.IF_ACMPEQ:
            case Const.IF_ACMPNE:
              break;

              // These instructions compare the integer on the top of the stack
              // to zero. There are no literals here (except 0)
            case Const.IFEQ:
            case Const.IFNE:
            case Const.IFLT:
            case Const.IFGE:
            case Const.IFGT:
            case Const.IFLE:
              {
                break;
              }

              // InstanceOf pushes either 0 or 1 on the stack depending on
              // whether
              // the object on top of stack is of the specified type.
              // If were interested in class literals, this would be interesting
            case Const.INSTANCEOF:
              break;

              // Duplicates the item on the top of stack. No literal.
            case Const.DUP:
              {
                break;
              }

              // Duplicates the item on the top of the stack and inserts it 2
              // values down in the stack. No literals
            case Const.DUP_X1:
              {
                break;
              }

              // Duplicates either the top 2 category 1 values or a single
              // category 2 value and inserts it 2 or 3 values down on the
              // stack.
            case Const.DUP2_X1:
              {
                break;
              }

              // Duplicate either one category 2 value or two category 1 values.
            case Const.DUP2:
              {
                break;
              }

              // Dup the category 1 value on the top of the stack and insert it
              // either
              // two or three values down on the stack.
            case Const.DUP_X2:
              {
                break;
              }

            case Const.DUP2_X2:
              {
                break;
              }

              // Pop instructions discard the top of the stack.
            case Const.POP:
              {
                break;
              }

              // Pops either the top 2 category 1 values or a single category 2
              // value
              // from the top of the stack.
            case Const.POP2:
              {
                break;
              }

              // Swaps the two category 1 types on the top of the stack.
            case Const.SWAP:
              {
                break;
              }

              // Compares two integers on the stack
            case Const.IF_ICMPEQ:
            case Const.IF_ICMPGE:
            case Const.IF_ICMPGT:
            case Const.IF_ICMPLE:
            case Const.IF_ICMPLT:
            case Const.IF_ICMPNE:
              {
                break;
              }

              // Get the value of a field
            case Const.GETFIELD:
              {
                break;
              }

              // stores the top of stack into a field
            case Const.PUTFIELD:
              {
                break;
              }

              // Pushes the value of a static field on the stack
            case Const.GETSTATIC:
              {
                break;
              }

              // Pops a value off of the stack into a static field
            case Const.PUTSTATIC:
              {
                break;
              }

              // pushes a local onto the stack
            case Const.DLOAD:
            case Const.DLOAD_0:
            case Const.DLOAD_1:
            case Const.DLOAD_2:
            case Const.DLOAD_3:
            case Const.FLOAD:
            case Const.FLOAD_0:
            case Const.FLOAD_1:
            case Const.FLOAD_2:
            case Const.FLOAD_3:
            case Const.ILOAD:
            case Const.ILOAD_0:
            case Const.ILOAD_1:
            case Const.ILOAD_2:
            case Const.ILOAD_3:
            case Const.LLOAD:
            case Const.LLOAD_0:
            case Const.LLOAD_1:
            case Const.LLOAD_2:
            case Const.LLOAD_3:
              {
                break;
              }

              // Pops a value off of the stack into a local
            case Const.DSTORE:
            case Const.DSTORE_0:
            case Const.DSTORE_1:
            case Const.DSTORE_2:
            case Const.DSTORE_3:
            case Const.FSTORE:
            case Const.FSTORE_0:
            case Const.FSTORE_1:
            case Const.FSTORE_2:
            case Const.FSTORE_3:
            case Const.ISTORE:
            case Const.ISTORE_0:
            case Const.ISTORE_1:
            case Const.ISTORE_2:
            case Const.ISTORE_3:
            case Const.LSTORE:
            case Const.LSTORE_0:
            case Const.LSTORE_1:
            case Const.LSTORE_2:
            case Const.LSTORE_3:
              {
                break;
              }

              // Push a value from the runtime constant pool. We'll get these
              // values when processing the constant pool itself
            case Const.LDC:
            case Const.LDC_W:
            case Const.LDC2_W:
              {
                break;
              }

              // Push the length of an array on the stack
            case Const.ARRAYLENGTH:
              {
                break;
              }

              // Push small constants (-1..5) on the stack.
            case Const.DCONST_0:
              {
                result.doubles.add(Double.valueOf(0));
                break;
              }
            case Const.DCONST_1:
              {
                result.doubles.add(Double.valueOf(1));
                break;
              }
            case Const.FCONST_0:
              {
                result.floats.add(Float.valueOf(0));
                break;
              }
            case Const.FCONST_1:
              {
                result.floats.add(Float.valueOf(1));
                break;
              }
            case Const.FCONST_2:
              {
                result.floats.add(Float.valueOf(2));
                break;
              }
            case Const.ICONST_0:
              {
                result.ints.add(Integer.valueOf(0));
                break;
              }
            case Const.ICONST_1:
              {
                result.ints.add(Integer.valueOf(1));
                break;
              }
            case Const.ICONST_2:
              {
                result.ints.add(Integer.valueOf(2));
                break;
              }
            case Const.ICONST_3:
              {
                result.ints.add(Integer.valueOf(3));
                break;
              }
            case Const.ICONST_4:
              {
                result.ints.add(Integer.valueOf(4));
                break;
              }
            case Const.ICONST_5:
              {
                result.ints.add(Integer.valueOf(5));
                break;
              }
            case Const.ICONST_M1:
              {
                result.ints.add(Integer.valueOf(-1));
                break;
              }
            case Const.LCONST_0:
              {
                result.longs.add(Long.valueOf(0));
                break;
              }
            case Const.LCONST_1:
              {
                result.longs.add(Long.valueOf(1));
                break;
              }

            case Const.BIPUSH:
            case Const.SIPUSH:
              {
                ConstantPushInstruction cpi = (ConstantPushInstruction) inst;
                result.ints.add((Integer) cpi.getValue());
                break;
              }

              // Primitive Binary operators.
            case Const.DADD:
            case Const.DCMPG:
            case Const.DCMPL:
            case Const.DDIV:
            case Const.DMUL:
            case Const.DREM:
            case Const.DSUB:
            case Const.FADD:
            case Const.FCMPG:
            case Const.FCMPL:
            case Const.FDIV:
            case Const.FMUL:
            case Const.FREM:
            case Const.FSUB:
            case Const.IADD:
            case Const.IAND:
            case Const.IDIV:
            case Const.IMUL:
            case Const.IOR:
            case Const.IREM:
            case Const.ISHL:
            case Const.ISHR:
            case Const.ISUB:
            case Const.IUSHR:
            case Const.IXOR:
            case Const.LADD:
            case Const.LAND:
            case Const.LCMP:
            case Const.LDIV:
            case Const.LMUL:
            case Const.LOR:
            case Const.LREM:
            case Const.LSHL:
            case Const.LSHR:
            case Const.LSUB:
            case Const.LUSHR:
            case Const.LXOR:
              break;

            case Const.LOOKUPSWITCH:
            case Const.TABLESWITCH:
              break;

            case Const.ANEWARRAY:
            case Const.NEWARRAY:
              {
                break;
              }

            case Const.MULTIANEWARRAY:
              {
                break;
              }

              // push the value at an index in an array
            case Const.AALOAD:
            case Const.BALOAD:
            case Const.CALOAD:
            case Const.DALOAD:
            case Const.FALOAD:
            case Const.IALOAD:
            case Const.LALOAD:
            case Const.SALOAD:
              {
                break;
              }

              // Pop the top of stack into an array location
            case Const.AASTORE:
            case Const.BASTORE:
            case Const.CASTORE:
            case Const.DASTORE:
            case Const.FASTORE:
            case Const.IASTORE:
            case Const.LASTORE:
            case Const.SASTORE:
              break;

            case Const.ARETURN:
            case Const.DRETURN:
            case Const.FRETURN:
            case Const.IRETURN:
            case Const.LRETURN:
            case Const.RETURN:
              {
                break;
              }

              // subroutine calls.
            case Const.INVOKESTATIC:
            case Const.INVOKEVIRTUAL:
            case Const.INVOKESPECIAL:
            case Const.INVOKEINTERFACE:
            case Const.INVOKEDYNAMIC:
              break;

              // Throws an exception.
            case Const.ATHROW:
              break;

              // Opcodes that don't need any modifications. Here for reference
            case Const.ACONST_NULL:
            case Const.ALOAD:
            case Const.ALOAD_0:
            case Const.ALOAD_1:
            case Const.ALOAD_2:
            case Const.ALOAD_3:
            case Const.ASTORE:
            case Const.ASTORE_0:
            case Const.ASTORE_1:
            case Const.ASTORE_2:
            case Const.ASTORE_3:
            case Const.CHECKCAST:
            case Const.D2F: // double to float
            case Const.D2I: // double to integer
            case Const.D2L: // double to long
            case Const.DNEG: // Negate double on top of stack
            case Const.F2D: // float to double
            case Const.F2I: // float to integer
            case Const.F2L: // float to long
            case Const.FNEG: // Negate float on top of stack
            case Const.GOTO:
            case Const.GOTO_W:
            case Const.I2B: // integer to byte
            case Const.I2C: // integer to char
            case Const.I2D: // integer to double
            case Const.I2F: // integer to float
            case Const.I2L: // integer to long
            case Const.I2S: // integer to short
            case Const.IFNONNULL:
            case Const.IFNULL:
            case Const.IINC: // increment local variable by a constant
            case Const.INEG: // negate integer on top of stack
            case Const.JSR: // pushes return address on the stack,
            case Const.JSR_W:
            case Const.L2D: // long to double
            case Const.L2F: // long to float
            case Const.L2I: // long to int
            case Const.LNEG: // negate long on top of stack
            case Const.MONITORENTER:
            case Const.MONITOREXIT:
            case Const.NEW:
            case Const.NOP:
            case Const.RET: // this is the internal JSR return
            case Const.WIDE:
              break;

              // Make sure we didn't miss anything
            default:
              throw new Error("instruction " + inst + " unsupported");
          }
        }
      }
    }
    return result;
  }

  /**
   * Convert a collection of ConstantSets to the format expected by GenTest.addClassLiterals.
   *
   * @param constantSets the sets of constantSets
   * @return a map of types to constant operations
   */
  public static MultiMap<Class<?>, NonreceiverTerm> toMap(Collection<ConstantSet> constantSets) {
    final MultiMap<Class<?>, NonreceiverTerm> map = new MultiMap<>();
    for (ConstantSet cs : constantSets) {
      Class<?> clazz;
      try {
        clazz = TypeNames.getTypeForName(cs.classname);
      } catch (ClassNotFoundException | NoClassDefFoundError e) {
        throw new Error("Class " + cs.classname + " not found on the classpath.");
      }
      for (Integer x : cs.ints) {
        map.add(clazz, new NonreceiverTerm(JavaTypes.INT_TYPE, x));
      }
      for (Long x : cs.longs) {
        map.add(clazz, new NonreceiverTerm(JavaTypes.LONG_TYPE, x));
      }
      for (Float x : cs.floats) {
        map.add(clazz, new NonreceiverTerm(JavaTypes.FLOAT_TYPE, x));
      }
      for (Double x : cs.doubles) {
        map.add(clazz, new NonreceiverTerm(JavaTypes.DOUBLE_TYPE, x));
      }
      for (String x : cs.strings) {
        map.add(clazz, new NonreceiverTerm(JavaTypes.STRING_TYPE, x));
      }
      for (Class<?> x : cs.classes) {
        map.add(clazz, new NonreceiverTerm(JavaTypes.CLASS_TYPE, x));
      }
    }
    return map;
  }
}
