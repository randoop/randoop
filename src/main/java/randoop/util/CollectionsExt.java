package randoop.util;

import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import org.plumelib.util.CollectionsPlume;
import org.plumelib.util.StringsPlume;
import randoop.Globals;

public final class CollectionsExt {

  private CollectionsExt() {
    throw new IllegalStateException("no instances");
  }

  static <T> T getNthIteratedElement(Collection<? extends T> s, int index) {
    if (s == null) throw new IllegalArgumentException("s cannot be null.");
    if (s.isEmpty()) throw new IllegalArgumentException("s cannot be empty.");
    if (index >= s.size()) {
      throw new IllegalArgumentException(
          "Index " + index + " out of bounds for set of size " + s.size());
    }
    return getNthIteratedElement(s.iterator(), index);
  }

  private static <T> T getNthIteratedElement(Iterator<? extends T> iter, int index) {
    if (index < 0) throw new IllegalArgumentException("Index " + index + " out of bounds");
    int counter = 0;
    while (iter.hasNext()) {
      if (counter == index) {
        return iter.next();
      }
      iter.next();
      counter++;
    }
    throw new IllegalArgumentException("index " + index + " out of bounds, size=" + counter);
  }

  /**
   * Returns the String.valueOf() of all elements of the collection, one on each line.
   *
   * @param c the collection of objects to include in string
   * @return the concatenated string of object strings as lines
   */
  static String toStringInLines(Collection<?> c) {
    if (c.isEmpty()) {
      return "";
    }
    return StringsPlume.joinLines(toStringLines(c)) + Globals.lineSep;
  }

  /**
   * List of String.valueOf() of all elements of the collection.
   *
   * @param c the collection of objects to include in string
   * @return the concatenated string of object strings
   */
  private static List<String> toStringLines(Collection<?> c) {
    return CollectionsPlume.mapList(String::valueOf, c);
  }
}
