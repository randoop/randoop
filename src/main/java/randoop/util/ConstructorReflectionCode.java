package randoop.util;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;

/** Wraps a constructor together with its arguments, ready for execution. Can be run only once. */
public final class ConstructorReflectionCode extends ReflectionCode {

  /** The constructor to be called. */
  private final Constructor<?> constructor;
  /**
   * The arguments that the constructor is applied to. If an inner class constructor has a receiver,
   * it is the first element of this array.
   */
  private final Object[] inputs;

  /**
   * Create a new ConstructorReflectionCode to represent a constructor invocation.
   *
   * @param constructor the constructor to be called
   * @param inputs the arguments that the constructor is applied to. If an inner class constructor
   *     has a receiver, it is the first element of this array.
   */
  @SuppressWarnings("deprecation") // AccessibleObject.isAccessible() has no replacement in Java 8.
  public ConstructorReflectionCode(Constructor<?> constructor, Object[] inputs) {
    if (constructor == null) {
      throw new IllegalArgumentException("constructor is null");
    }
    if (inputs == null) {
      throw new IllegalArgumentException("inputs is null");
    }
    this.constructor = constructor;
    this.inputs = inputs;

    if (!this.constructor.isAccessible()) {
      this.constructor.setAccessible(true);
      Log.logPrintf("not accessible: %s%n", this.constructor);
      // TODO something is bizarre - it seems that a public method can be
      // not-accessible sometimes. RatNum(int,int)
      // TODO you cannot just throw the exception below - because no sequences
      // will be created in the randoop.experiments.
      // throw new IllegalStateException("Not accessible: " + this.constructor);
    }
  }

  @SuppressWarnings("Finally")
  @Override
  public void runReflectionCodeRaw() {
    try {
      this.retval = this.constructor.newInstance(this.inputs);
    } catch (InvocationTargetException e) {
      // The underlying constructor threw an exception
      this.exceptionThrown = e.getCause();
    } catch (Throwable e) {
      // Any other exception indicates Randoop should not have called the constructor
      throw new ReflectionCodeException(e);
    }
  }

  @Override
  public String toString() {
    return "Call to " + constructor + ", args: " + Arrays.toString(inputs) + status();
  }
}
