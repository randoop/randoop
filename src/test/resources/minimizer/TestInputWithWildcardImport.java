import java.util.*;
import org.junit.FixMethodOrder;
import org.junit.Test;
import org.junit.runners.MethodSorters;

@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestInputWithWildcardImport {

  // Test input with wildcard import already included. Redundant import statements will not be
  // added by the minimizer.
  @Test
  public void test1() throws Throwable {
    java.util.List<java.lang.Integer> list = new java.util.ArrayList<java.lang.Integer>();
    list.add(1);
    list.add(2);

    org.junit.Assert.assertTrue(list.size() == 2);

    list.add(3);
    // False, should be 3.
    org.junit.Assert.assertTrue(list.size() == 2);
  }
}
