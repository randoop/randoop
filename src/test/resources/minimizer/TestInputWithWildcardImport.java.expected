import java.util.*;
import org.junit.FixMethodOrder;
import org.junit.Test;
import org.junit.runners.MethodSorters;

@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestInputWithWildcardImportMinimized {

    @Test
    public void test1() throws Throwable {
        List<java.lang.Integer> list = new ArrayList<java.lang.Integer>();
        list.add(1);
        list.add(2);
        list.add(3);
        // False, should be 3.
        org.junit.Assert.assertTrue(list.size() == 2);
    }
}
